// __BEGIN_LICENSE__
// Copyright (C) 2006-2010 United States Government as represented by
// the Administrator of the National Aeronautics and Space Administration.
// All Rights Reserved.
// __END_LICENSE__


package vw.platefile;
import "vw/Plate/Rpc.proto";
import "vw/Plate/IndexData.proto";
import "vw/Plate/IndexDataPrivate.proto";

// ----------------------------------
// Client --> Server Requests
// ----------------------------------

message IndexOpenRequest {
  required string plate_name = 1;
}

message IndexCreateRequest {
  required string plate_name = 1;
  required IndexHeader index_header = 2;
}

// If given, the arguments act as filters
message IndexListRequest {
  optional string tile_filetype = 1; // file type [e.g. tif, png, or jpg]
  optional int32 pixel_format   = 2; // see vw/Image/PixelTypeInfo.h
  optional int32 channel_type   = 3; // see vw/Image/PixelTypeInfo.h
  optional string type          = 4; // platefile type [ toast, kml, or gigapan ]
}

message IndexInfoRequest {
  required int32 platefile_id = 1;
}

message IndexLogRequest {
  required int32 platefile_id = 1;
  required string message = 2;
}

message IndexReadRequest {
  required int32 platefile_id = 1;
  required int32 col = 3;
  required int32 row = 4;
  required int32 level = 5;
  required int32 transaction_id = 6;
  required bool exact_transaction_match = 7;
}

message IndexPageRequest {
  required int32 platefile_id = 1;
  required int32 col = 3;
  required int32 row = 4;
  required int32 level = 5;
}

message IndexWriteRequest {
  required int32 platefile_id = 1;
}

message IndexWriteUpdate {
  required int32 platefile_id = 1;
  required TileHeader header = 3;
  required detail.IndexRecord record = 4;
}

message IndexMultiWriteUpdate {
  repeated IndexWriteUpdate write_updates = 2;
}

message IndexWriteComplete {
  required int32 platefile_id = 1;
  required uint32 blob_id = 2;
  required uint64 blob_offset = 3;
}

message IndexTransactionRequest {
  required int32 platefile_id = 1;
  required string description = 2;
  optional int32 transaction_id_override = 3 [default = -1];
}

message IndexTransactionComplete {
  required int32 platefile_id = 1;
  required uint32 transaction_id = 2;
  required bool update_read_cursor = 3;
}

message IndexTransactionFailed {
  required int32 platefile_id = 1;
  required uint32 transaction_id = 2;
}

message IndexTransactionCursorRequest {
  required int32 platefile_id = 1;
}

message IndexNumLevelsRequest {
  required int32 platefile_id = 1;
}

message IndexTestRequest {
  required uint64 value = 1;
}

// ----------------------------------
// Server --> Client Responses
// ----------------------------------

message IndexOpenReply {
  required string short_plate_filename = 1;
  required string full_plate_filename = 2;
  required IndexHeader index_header = 3;
}

message IndexInfoReply {
  required string short_plate_filename = 1;
  required string full_plate_filename = 2;
  required IndexHeader index_header = 3;
}

message IndexListReply {
  repeated string platefile_names = 1;
}

message IndexPageReply {
  required bytes page_bytes = 1;
}

message IndexReadReply {
  required detail.IndexRecord index_record = 1;
}

message IndexSuccess {
  optional string message = 1;
}

message IndexWriteReply {
  required uint32 blob_id = 1;
  required uint64 size = 2;
}

message IndexTransactionReply {
  required uint32 transaction_id = 1;
}

message IndexTransactionCursorReply {
  required uint32 transaction_id = 1;
}

message IndexNumLevelsReply {
  required int32 num_levels = 1;
}

message IndexTestReply {
  required uint64 value = 1;
}

option cc_generic_services = true;

service IndexService {

  // Basic Platefile Management
  rpc OpenRequest (IndexOpenRequest) returns (IndexOpenReply);
  rpc CreateRequest (IndexCreateRequest) returns (IndexOpenReply);
  rpc InfoRequest (IndexInfoRequest) returns (IndexInfoReply);

  // Platefile I/O
  rpc PageRequest (IndexPageRequest) returns (IndexPageReply);
  rpc ReadRequest (IndexReadRequest) returns (IndexReadReply);
  rpc WriteRequest (IndexWriteRequest) returns (IndexWriteReply);
  rpc WriteUpdate (IndexWriteUpdate) returns (RpcNullMsg);
  rpc MultiWriteUpdate (IndexMultiWriteUpdate) returns (RpcNullMsg);
  rpc WriteComplete (IndexWriteComplete) returns (RpcNullMsg);


  // Transaction Management
  rpc TransactionRequest (IndexTransactionRequest) returns (IndexTransactionReply);
  rpc TransactionComplete (IndexTransactionComplete) returns (RpcNullMsg);
  rpc TransactionFailed (IndexTransactionFailed) returns (RpcNullMsg);
  rpc TransactionCursor (IndexTransactionCursorRequest) returns (IndexTransactionCursorReply);

  // Misc.
  rpc NumLevelsRequest (IndexNumLevelsRequest) returns (IndexNumLevelsReply);
  rpc ListRequest (IndexListRequest) returns (IndexListReply);
  rpc TestRequest (IndexTestRequest) returns (IndexTestReply);
  rpc LogRequest (IndexLogRequest) returns (RpcNullMsg);

}
